python/pycrdt/_awareness.py

```diff
@@ -38,164 +36,148 @@ def states(self) -> dict[int, dict[str, Any]]:
-            client_msg.insert(0, write_var_uint(meta.get("clock", 0)))
-            client_msg.insert(0, write_var_uint(client_id))
-            messages.append(b"".join(client_msg))
+            state = self._states.get(client_id)
+            clock = cast(int, self._meta.get(client_id, {}).get("clock"))
+            encoder.write_var_uint(client_id)
+            encoder.write_var_uint(clock)
+            encoder.write_var_string(json.dumps(state, separators=(",", ":")))
+        return encoder.to_bytes()
 
-        if not messages:
-            return None
-
-        messages.insert(0, write_var_uint(len(client_ids)))
-        encoded_messages = b"".join(messages)
+    def apply_awareness_update(self, update: bytes, origin: Any) -> None:
+        """
+        Applies the binary update and notifies subscribers with changes.
 
-        message = [
-            write_var_uint(YMessageType.AWARENESS),
-            write_var_uint(len(encoded_messages)),
-            encoded_messages,
-        ]
-        return b"".join(message)
+        Args:
+            update: The binary update.
+            origin: The origin of the update.
+        """
+        decoder = Decoder(update)
+        timestamp = int(time.time() * 1000)
+        added = []
+        updated = []
+        filtered_updated = []
+        removed = []
+        length = decoder.read_var_uint()
+        for _ in range(length):
+            client_id = decoder.read_var_uint()
+            clock = decoder.read_var_uint()
+            state_str = decoder.read_var_string()
+            state = None if not state_str else json.loads(state_str)
+            client_meta = self._meta.get(client_id)
+            prev_state = self._states.get(client_id)
+            curr_clock = 0 if client_meta is None else client_meta["clock"]
+            if curr_clock < clock or (
+                curr_clock == clock and state is None and client_id in self._states
+            ):
+                if state is None:
+                    # Never let a remote client remove this local state.
+                    if client_id == self.client_id and self.get_local_state() is not None:
+                        # Remote client removed the local state. Do not remove state.
+                        # Broadcast a message indicating that this client still exists by increasing
+                        # the clock.
+                        clock += 1
+                    else:
+                        if client_id in self._states:
+                            del self._states[client_id]
+                else:
+                    self._states[client_id] = state
+                self._meta[client_id] = {
+                    "clock": clock,
+                    "lastUpdated": timestamp,
+                }
+                if client_meta is None and state is not None:
+                    added.append(client_id)
+                elif client_meta is not None and state is None:
+                    removed.append(client_id)
```

Comment:
Will we not fallback here if a remote client try to remove the local state ?

Reply:
It's the translation of [this](https://github.com/yjs/y-protocols/blob/255b7e6c96d9c6dd48b5519b3f83ab65afdb4bba/awareness.js#L275-L276). Do you mean that there is an issue in the JavaScript implementation?

Reply:
Maybe...
We should test it, but it seems to me that if a remote client try to set the local awareness to null, this condition is fulfilled.
The `client_meta` is the local meta (and may be not null), and the `state` is the state to apply, which is null in that case.

Reply:
I agree with that, but it will send a wrong information to other client, that the local state has been removed, no ?